### PR TITLE
Add an assertTypes utility

### PR DESCRIPTION
**Pull request by samhorlbeck**
See the comment and the tests for usage.

There are unrelated prettier changes

### CODE REVIEW

File: packages-exp/auth-exp/src/core/util/assert.ts [hunk 1]
```diff
@@ -44,6 +44,92 @@ export function assert(
...
+ *
+ * @param args
+ * @param expected
+ */
+export function assertTypes(
+  args: Omit<IArguments, 'callee'>,
+  ...expected: Array<TypeExpectation | Optional>
+): void {
+  if (args.length > expected.length) {
+    fail('TODO', AuthErrorCode.ARGUMENT_ERROR);
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/util/assert.ts:89]: can this be debugFail? We should never hit this based on user input, right?

File: packages-exp/auth-exp/src/core/util/assert.ts [hunk 2]
```diff
@@ -44,6 +44,92 @@ export function assert(
...
+      const record = arg as Record<string, unknown>;
+      const map = expect as MapType;
+      const keys = Object.keys(expect);
+
+      assertTypes(
+        keys.map(k => record[k]),
+        ...keys.map(k => map[k])
+      );
+    } else {
+      assert(arg instanceof expect, 'app', AuthErrorCode.ARGUMENT_ERROR);
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/util/assert.ts:128]: same here, we should probably use debugAssert and not require an app

File: packages-exp/auth-exp/src/core/util/assert.ts [hunk 3]
```diff
@@ -44,6 +44,92 @@ export function assert(
...
+        'TODO',
+        AuthErrorCode.ARGUMENT_ERROR
+      );
+    } else if (typeof expect === 'object') {
+      // Recursively check record arguments
+      const record = arg as Record<string, unknown>;
+      const map = expect as MapType;
+      const keys = Object.keys(expect);
+
+      assertTypes(
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/util/assert.ts:123]: will the recursion here work with cross referenced types like Auth & User that point at each other via this.auth and this.currentUser?

File: packages-exp/auth-exp/src/core/util/assert.ts (hunk 1 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/assert.ts:89]: Nope, for example:

```ts
function foo(auth: Auth) {
  assertTypes(arguments, AuthImpl);
}
```
should fail with this call (which is perfectly valid in javascript)
```js
foo(auth, 'some-other-parameter');
```

File: packages-exp/auth-exp/src/core/util/assert.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/assert.ts:128]: These are runtime checks and _should_ fail in the production build

File: packages-exp/auth-exp/src/core/util/assert.ts (hunk 3 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/assert.ts:123]: The recursion only happens for record types (not for instance checks). So
```ts
function foo(profile: {displayName?: string}) {
  assertTypes(arguments, {displayName: opt(string)});
}
```
will recurse, whereas
```
function foo(auth: Auth) {
  assertTypes(arguments, AuthImpl);
}
```
will not

**Pull request merged**